src/Hedgehog/Extras/Test/New/Golden.hs

```diff
@@ -0,0 +1,199 @@
+import           Hedgehog (MonadTest)
+import           Hedgehog.Extras.Test.Base (failMessage)
+import           System.Directory (canonicalizePath, getCurrentDirectory)
+import           System.FilePath (takeDirectory, makeRelative)
+import           System.IO (IO, FilePath)
+import           System.IO.Unsafe (unsafePerformIO)
+import qualified System.IO as IO
+
+import qualified Control.Concurrent.QSem as IO
+import qualified Data.List as List
+import qualified Data.Map as Map
+import qualified GHC.Stack as GHC
+import qualified Hedgehog.Extras.Test as H
+import qualified Hedgehog.Internal.Property as H
+import qualified System.Directory as IO
+import qualified System.Environment as IO
+import qualified System.IO.Unsafe as IO
+
+semBracket :: ()
+  => MonadIO m
+  => MonadMask m
+  => IO.QSem
+  -> m a
+  -> m a
+semBracket sem =
+  bracket_
+    (liftIO (IO.waitQSem sem))
+    (liftIO (IO.signalQSem sem))
+
+-- | The file to log whenever a golden file is referenced.
+mGoldenFileLogFile :: Maybe FilePath
+mGoldenFileLogFile = IO.unsafePerformIO $
+  IO.lookupEnv "GOLDEN_FILE_LOG_FILE"
+
+-- | Whether the test should create the golden files if the files do not exist.
+createGoldenFiles :: Bool
+createGoldenFiles = IO.unsafePerformIO $ do
+  value <- IO.lookupEnv "CREATE_GOLDEN_FILES"
+  return $ value == Just "1"
+
+-- | Whether the test should recreate the golden files if the files already exist.
+recreateGoldenFiles :: Bool
+recreateGoldenFiles = IO.unsafePerformIO $ do
+  value <- IO.lookupEnv "RECREATE_GOLDEN_FILES"
+  return $ value == Just "1"
+
+writeGoldenFile :: ()
+  => HasCallStack
+  => MonadIO m
+  => MonadTest m
+  => FilePath
+  -> String
+  -> m ()
+writeGoldenFile goldenFile actualContent = GHC.withFrozenCallStack $ do
+  H.note_ $ "Creating golden file " <> goldenFile
+  H.createDirectoryIfMissing_ (takeDirectory goldenFile)
+  H.writeFile goldenFile actualContent
+
+reportGoldenFileMissing :: ()
+  => HasCallStack
+  => MonadIO m
+  => MonadTest m
+  => FilePath
+  -> m ()
+reportGoldenFileMissing goldenFile = GHC.withFrozenCallStack $ do
+  H.note_ $ unlines
+    [ "Golden file " <> goldenFile <> " does not exist."
+    , "To create it, run with CREATE_GOLDEN_FILES=1."
+    , "To recreate it, run with RECREATE_GOLDEN_FILES=1."
+    ]
+  H.failure
+
+checkAgainstGoldenFile :: ()
+  => HasCallStack
+  => MonadIO m
+  => MonadTest m
+  => FilePath
+  -> [String]
+  -> m ()
+checkAgainstGoldenFile goldenFile actualLines = GHC.withFrozenCallStack $ do
+  referenceLines <- List.lines <$> H.readFile goldenFile
+  let difference = getGroupedDiff actualLines referenceLines
+  case difference of
+    []       -> pure ()
+    [Both{}] -> pure ()
+    _        -> do
+      H.note_ $ unlines
+        [ "Golden test failed against the golden file."
+        , "To recreate golden file, run with RECREATE_GOLDEN_FILES=1."
+        ]
+      failMessage callStack $ ppDiff difference
+
+tvGoldenFileSems :: STM.TVar (Map FilePath IO.QSem)
```

Comment:
There still will be contention on the semaphore map access. Not that big as in the previous version with a single semaphore. Why not use: https://hackage-content.haskell.org/package/stm-containers-1.2.1.1/docs/StmContainers-Map.html

https://nikita-volkov.github.io/stm-containers/

>The “stm-containers” scale well up to 12 threads. This is quite expected for a 6-core hyperthreaded processor, which the benchmark has been run on.
>TVar (HashMap (TVar a)) basically stops scaling after 2-4 threads (...)

I'm using 32 threads to run tests personally 😄

Reply:
We can optimise it later of course.  Any optimisations will be local to the file and not change the API.